RavenDB-19331: Implementation of key reuse scopes & other key encoding optimizations

redknightlois opened the pull request:
This is still a work in process. It is still not optimized, just looking to ensure it is correct.

redknightlois committed:
RavenDB-19331: Improvements over several code-paths.

redknightlois committed:
RavenDB-19331: Much faster `.Set()` because we avoid the call instruction.

redknightlois committed:
RavenDB-19331: Cleaning up the code.

redknightlois commented:
Trying to figure out if I can get rid entirely of the hashing structure, that will provide much better memory locality and compact code.

redknightlois committed:
RavenDB-19331: Removed altogether the hashing method.

redknightlois committed:
RavenDB-19331: Improvements on key creators and accessors.

redknightlois committed:
RavenDB-19331: Incorrect check causes to invalidate cache every iteration.

redknightlois committed:
RavenDB-19331: Last value was not updated, therefore it didn't play any role.

redknightlois committed:
RavenDB-19331: Skipping over a Variable Size Encoding is faster than decoding.

redknightlois committed:
RavenDB-19331: Set optimizations

ayende commented on src/Sparrow/Compression/VariableSizeEncoding.cs:
```diff
@@ -261,6 +261,51 @@ private static void ThrowNotEnoughOutputSpace()
+            {
+                int bytesReadWhenOverflow;
+                if (typeof(T) == typeof(long) || typeof(T) == typeof(ulong))
+                    bytesReadWhenOverflow = 12;
```
variable len size of long is 9, for ulong, 10

redknightlois replied:
No.

![image](https://user-images.githubusercontent.com/1045411/198328209-82e70613-6015-44cb-becb-6f0fdf9cd241.png)

ayende commented on src/Sparrow/Compression/VariableSizeEncoding.cs:
```diff
@@ -261,6 +261,51 @@ private static void ThrowNotEnoughOutputSpace()
+                if (typeof(T) == typeof(long) || typeof(T) == typeof(ulong))
+                    bytesReadWhenOverflow = 12;
+                else if (typeof(T) == typeof(int) || typeof(T) == typeof(uint))
+                    bytesReadWhenOverflow = 7;
```
For all of those, that should be 5, no?

redknightlois replied:
They have to be the same as the other. They are all wrong by 2.

ayende commented on src/Voron/Data/Sets/Set.cs:
```diff
@@ -630,10 +639,10 @@ public Iterator(Set parent)
 
-                ref var state = ref _parent._stk[_parent._pos];
-                var leafPage = new SetLeafPage(state.Page);
+                //ref var state = ref _parent._stk[_parent._pos];
```
can remove this?

ayende commented on src/Voron/Data/CompactTrees/CompactTree.cs:
```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
+            private const int MappingTableSize = 64;
+            private const int MappingTableMask = MappingTableSize - 1;
+
+            private readonly long[] _keyMappingDictionary = new long[MappingTableSize];
+            private readonly int[] _keyMappingIndex = new int[MappingTableSize];
```
Those are used only _during_ the searching process, right?
Can we make the thread static, then? Would save some allocations.

redknightlois replied:
Yes. Haven't look yet at what allocation avoidance technique would be proper to use here.

ayende commented on src/Voron/Data/CompactTrees/CompactTree.cs:
```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
+                    // IMPORTANT: Pointers are potentially invalidated by the grow storage call but not the indexes.
+
+                    // We look for an appropriate place to put this key
+                    int buckedIdx = SelectHashBucketForWrite();
```
I though we aren't doing hashing any more?

redknightlois replied:
Need to change the name, yes.

ayende commented on src/Voron/Data/CompactTrees/CompactTree.cs:
```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
+            private int SelectHashBucketForWrite()
             {
-                var dictionary = tree.GetEncodingDictionary(dictionaryId);
+                _keyMappingCurrent++;
```
this is linear, so great, but we need a better name here, I think

arekpalinski commented:
@redknightlois is it still draft or something that we could already merge?  (after we merge https://github.com/ravendb/ravendb/pull/15118 and confirm that everything works). Or you want to test it on real data before doing merge?

redknightlois commented:
While still working on this, and therefore not ready for merging, would be
great if we can give it a run on a real-world. Working on optimizing the
key comparisons so when that is done and the results are I will ponder if
merging it as an intermediate step.

On Thu, Oct 20, 2022, 4:39 AM Arkadiusz Paliński ***@***.***>
wrote:

> @redknightlois <https://github.com/redknightlois> is it still draft or
> something that we could already merge? (after we merge #15118
> <https://github.com/ravendb/ravendb/pull/15118> and confirm that
> everything works). Or you want to test it on real data before doing merge?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/pull/15071#issuecomment-1285080920>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH7HI5L4C5QXMXSYHM6VCLWEDZJTANCNFSM6AAAAAAQ7GMXXM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

redknightlois committed:
RavenDB-19331: Initial implementation of memory compare with updates for newer runtimes.

redknightlois committed:
RavenDB-19331: Explicit flow control (276 bytes)

|        Method |  Length |          Mean |       Error |      StdDev | Code Size |
|-------------- |-------- |--------------:|------------:|------------:|----------:|
| Avx2_Expanded |      64 |      1.217 ns |   0.0474 ns |   0.0679 ns |     276 B |
|   SpanHelpers |      64 |      2.361 ns |   0.0571 ns |   0.0534 ns |     417 B |
| Avx2_Expanded |     255 |      4.614 ns |   0.1022 ns |   0.0956 ns |     276 B |
|   SpanHelpers |     255 |      4.039 ns |   0.0846 ns |   0.0791 ns |     417 B |
| Avx2_Expanded |     256 |      4.817 ns |   0.0583 ns |   0.0545 ns |     276 B |
|   SpanHelpers |     256 |      3.997 ns |   0.1016 ns |   0.1356 ns |     417 B |
| Avx2_Expanded |  262144 |  4,327.549 ns |  75.1124 ns |  66.5852 ns |     276 B |
|   SpanHelpers |  262144 |  4,359.495 ns |  83.3043 ns |  81.8160 ns |     417 B |
| Avx2_Expanded | 2097152 | 36,097.177 ns | 590.1075 ns | 551.9869 ns |     276 B |
|   SpanHelpers | 2097152 | 37,088.025 ns | 585.4390 ns | 547.6200 ns |     417 B |

redknightlois committed:
RavenDB-19331: Improved test harness.

redknightlois committed:
RavenDB-19331: Prefetch enabled version

redknightlois committed:
RavenDB-19331: Branch separated (271 bytes)

redknightlois committed:
RavenDB-19331: Forced memory aligned with prefetch (313 bytes)

redknightlois committed:
RavenDB-19331: Include newer compare routine.

arekpalinski commented:
Can you please fix the compilation and solve the conflict?

redknightlois committed:
RavenDB-19331: Implementation of key reuse scopes.

redknightlois committed:
RavenDB-19331: Optimizations for small size variable size encoding reads

redknightlois committed:
RavenDB-19331: Avoiding span creations for simple accesses

redknightlois committed:
RavenDB-19331: Fixed "Dictionary not set" error found on production databases.

redknightlois committed:
RavenDB-19331: Compare alignment fixes that impact performance profile.

redknightlois committed:
RavenDB-19331: Improvements into the decoding of the key.

redknightlois committed:
RavenDB-19331: Improvements to codegen of 3-gram decoder.

redknightlois committed:
RavenDB-19331: Addressing review comments and Jenkins refusing to compile AVX2 microbenchmark

redknightlois committed:
RavenDB-19331: Optimized allocations and some operations with AVX2 if available.

redknightlois committed:
RavenDB-19331: Ensure we always use first the closest cache spot on a release.

redknightlois committed:
RavenDB-19331: Precompute the nodes span to avoid recalculating it every time.

redknightlois committed:
RavenDB-19331: Improved codegen on inlineable PushPage

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1312,6 +1316,7 @@ private void InsertTextualField(Tree fieldsTree, IndexedField indexedField, Span
                         dumper.WriteRemoval(term, ttt);
                         break;
                 }
+                scope.Dispose();
```
this looks weird
I assume you aren't using `using` here because of the `try/finaly` cost?

redknightlois replied:
Yes. We are avoiding the `try/finally` but returning it if possible.

arekpalinski commented:
Can you please check this test https://jenkins.hibernatingrhinos.com/job/PR_Tests/3773/testReport/junit/SlowTests.Corax.Bugs/CompactTreeAddAndRemove/gz__/ ?

redknightlois committed:
RavenDB-19331: Improvements to GetEncodedEntry

redknightlois committed:
RavenDB-19331: Ensure we bail out as soon as we find the key.

redknightlois committed:
RavenDB-19331: Improve the codegen for several high traffic calls.


WIP


WIP

redknightlois committed:
RavenDB-19498: GetEncodedEntryBuffer reading keys with length=0
